core/src/com/unciv/logic/city/CityInfo.kt

```diff
@@ -175,60 +181,105 @@ class CityInfo {
-            .mapNotNull {
-                it.nation.cities
-                    .lastOrNull { city -> city !in usedCityNames }
-            }
-        if (newNames.isNotEmpty()) {
-            return newNames.random()
-        }
+    /**
+     * Borrows a city name from another major civilization.
+     *
+     * @param foundingCiv The civilization that founded this city.
+     * @param aliveCivs Every civilization currently alive.
+     * @param usedCityNames Every city name that have already been taken.
+     * @return A new city named in [String]. Null if failed to generate a name.
+     */
+    private fun borrowCityName(
+        foundingCiv: CivilizationInfo,
+        aliveCivs: Set<CivilizationInfo>,
+        usedCityNames: Set<String>
+    ): String? {
+        val aliveMajorNations: Sequence<Nation> =
+            aliveCivs.asSequence().filter { civ -> civ.isMajorCiv() }.map { civ -> civ.nation }
+
+        /*
+        We take the last unused city name for each other major nation in this game,
+        skipping nations whose names are exhausted,
+        and choose a random one from that pool if it's not empty.
+        */
+        val otherMajorNations: Sequence<Nation> =
+            aliveMajorNations.filter { nation -> nation != foundingCiv.nation }
+        var newCityNames: Set<String> =
+            otherMajorNations.mapNotNull { nation ->
+                nation.cities.lastOrNull { city -> city !in usedCityNames }
+            }.toSet()
+        if (newCityNames.isNotEmpty()) return newCityNames.random()
 
         // As per fandom wiki, once the names from the other nations in the game are exhausted,
-        // names are taken from the rest of the nations in the ruleset
-        newNames = getRuleset()
-            .nations
-            .filter { it.key !in civInfo.gameInfo.civilizations.map { civ -> civ.nation.name } }
-            .values
-            .map {
-                it.cities
-                    .filter { city -> city !in usedCityNames }
-            }.flatten()
-        if (newNames.isNotEmpty()) {
-            return newNames.random()
-        }
+        // names are taken from the rest of the major nations in the rule set
+        val absentMajorNations: Sequence<Nation> =
+            getRuleset().nations.values.asSequence().filter { nation ->
+                nation.isMajorCiv() && nation !in aliveMajorNations
+            }
+        newCityNames =
+            absentMajorNations.flatMap { nation ->
+                nation.cities.asSequence().filter { city -> city !in usedCityNames }
+            }.toSet()
+        if (newCityNames.isNotEmpty()) return newCityNames.random()
+
         // If for some reason we have used every single city name in the game,
-        // (are we using some sort of baserule mod without city names?)
-        // just return something so we at least have a name
-        return "The City without a Name"
+        // (are we using some sort of rule set mod without city names?)
```

Comment:
Very minor point, but iirc 'ruleset' is a single word

Reply:
I knew I cannot rely on Android Studio...

Reply:
Since this is a comment I'll fix this in one of the future PRs; I have a feeling that I'll come back to these methods really soon... Thanks a lot for pointing this out though; I definitely missed it without you.